// webdriver-image-comparison/build/clientSideScripts/getAndroidStatusAddressToolBarHeight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getAndroidStatusAddressToolBarHeight(androidOffsets, isHybridApp) {
    var _a = window.screen, height = _a.height, width = _a.width;
    var innerHeight = window.innerHeight;
    var match = (navigator.appVersion).match(/Android (\d+).?(\d+)?.?(\d+)?/);
    var majorVersion = parseInt(match[1], 10);
    var versionOffsets = androidOffsets[majorVersion];
    var statusAddressBarHeight = versionOffsets.STATUS_BAR + (isHybridApp ? 0 : versionOffsets.ADDRESS_BAR);
    var toolBarHeight = height - innerHeight - statusAddressBarHeight;
    if (toolBarHeight < 0) {
        toolBarHeight = versionOffsets.TOOL_BAR;
    }
    return {
        statusAddressBar: {
            height: statusAddressBarHeight,
            width: width,
            x: 0,
            y: 0,
        },
        toolBar: {
            height: toolBarHeight,
            width: width,
            x: 0,
            y: height - toolBarHeight,
        },
    };
}
exports.default = getAndroidStatusAddressToolBarHeight;
