// webdriver-image-comparison/build/helpers/constants.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_TABBABLE_OPTIONS = exports.OFFSETS = exports.FULL_PAGE_SCROLL_TIMEOUT = exports.DEFAULT_RESIZE_DIMENSIONS = exports.CUSTOM_CSS_ID = exports.DESKTOP = exports.DEFAULT_SHADOW = exports.FOLDERS = exports.PLATFORMS = exports.DEFAULT_FORMAT_STRING = void 0;
exports.DEFAULT_FORMAT_STRING = '{tag}-{browserName}-{width}x{height}-dpr-{dpr}';
exports.PLATFORMS = {
    ANDROID: 'android',
    IOS: 'ios',
};
exports.FOLDERS = {
    ACTUAL: 'actual',
    DIFF: 'diff',
    TEMP_FULL_SCREEN: 'tempFullScreen',
    DEFAULT: {
        BASE: './wic/baseline/',
        SCREENSHOTS: '.tmp/',
    }
};
exports.DEFAULT_SHADOW = {
    ADDRESS_BAR: 6,
    TOOL_BAR: 6,
};
exports.DESKTOP = 'desktop';
exports.CUSTOM_CSS_ID = 'pic-css';
exports.DEFAULT_RESIZE_DIMENSIONS = {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
};
exports.FULL_PAGE_SCROLL_TIMEOUT = 1500;
var androidDefaultOffsets = {
    STATUS_BAR: 24,
    ADDRESS_BAR: 56,
    TOOL_BAR: 48,
};
var iosDefaultOffsets = {
    STATUS_BAR: 20,
    STATUS_BAR_PRO: 24,
    STATUS_BAR_X: 44,
    ADDRESS_BAR: 50,
    HOME_BAR: {
        DEFAULT: {
            height: 5,
            width: 135,
            x: 120,
            y: 799,
        },
        LARGE: {
            height: 5,
            width: 148,
            x: 133,
            y: 883,
        },
    },
};
exports.OFFSETS = {
    ANDROID: {
        6: __assign({}, androidDefaultOffsets),
        7: __assign({}, androidDefaultOffsets),
        8: __assign({}, androidDefaultOffsets),
        9: __assign({}, androidDefaultOffsets),
        10: __assign({}, androidDefaultOffsets),
        11: __assign({}, androidDefaultOffsets),
        12: __assign({}, androidDefaultOffsets),
    },
    IOS: {
        10: {
            STATUS_BAR: 20,
            STATUS_BAR_PRO: 20,
            STATUS_BAR_X: 20,
            ADDRESS_BAR: 44,
            HOME_BAR: {
                DEFAULT: {
                    height: 0,
                    width: 0,
                    x: 0,
                    y: 0,
                },
                LARGE: {
                    height: 0,
                    width: 0,
                    x: 0,
                    y: 0,
                },
            },
        },
        11: __assign({}, iosDefaultOffsets),
        12: __assign({}, iosDefaultOffsets),
        13: __assign({}, iosDefaultOffsets),
        14: __assign({}, iosDefaultOffsets),
        15: __assign(__assign({}, iosDefaultOffsets), { ADDRESS_BAR: 0, STATUS_BAR_X: 47 }),
    },
};
exports.DEFAULT_TABBABLE_OPTIONS = {
    circle: {
        backgroundColor: '#ff0000',
        borderColor: '#000',
        borderWidth: 1,
        fontColor: '#fff',
        fontFamily: 'Arial',
        fontSize: 10,
        size: 10,
        showNumber: true,
    },
    line: {
        color: '#000',
        width: 1,
    },
};
